### PR TITLE
added hts filter allele variants

### PR DESCRIPTION
**Pull request by cseed**
moved tests to python

### CODE REVIEW

File: python/hail/methods/statgen.py [hunk 1]
```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
...
+          0 | 20
+          1 | 0    10
+            +-----------
+              0    1
+
+        In summary:
+
+        - GT: Downcode filtered alleles to reference.
+        - AD: The filtered alleles' columns are eliminated and their
+          value is added to the reference, e.g., filtering alleles 1
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: "Columns of filtered alleles are eliminated and their values are added to the reference column"

File: python/hail/methods/statgen.py [hunk 2]
```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
...
+               0      1
+
+        In summary:
+
+        - GT: Set to most likely genotype based on the PLs ignoring the filtered allele(s).
+        - AD: The filtered alleles' columns are eliminated, e.g., filtering alleles 1 and 2 transforms ``25,5,10,20`` to ``25,20``.
+        - DP: No change.
+        - PL: The filtered alleles' columns are eliminated and the remaining columns shifted so the minimum value is 0.
+        - GQ: The second-lowest PL (after shifting).
+
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: move newline below """

File: python/hail/methods/statgen.py [hunk 3]
```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
...
+        - AD: The filtered alleles' columns are eliminated and their
+          value is added to the reference, e.g., filtering alleles 1
+          and 2 transforms ``25,5,10,20`` to ``40,20``.
+        - DP: No change.
+        - PL: Downcode filtered alleles to reference, combine PLs
+          using minimum for each overloaded genotype, and shift so the
+          overall minimum PL is 0.
+        - GQ: The second-lowest PL (after shifting).
+        """
+        ds = self._ds
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: add newline above

File: python/hail/methods/statgen.py [hunk 4]
```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
...
+
+        **The downcode algorithm**
+
+        .. code-block:: text
+
+            GT: Call,
+            AD: Array[!Int32],
+            DP: Int32,
+            GQ: Int32,
+            PL: Array[!Int32].
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: period?

File: python/hail/methods/statgen.py [hunk 5]
```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
...
+        ds = self._ds
+        newPL = functions.cond(
+            functions.is_defined(ds.PL),
+            functions.range(0, self.new_v.num_genotypes()).map(
+                lambda newi: (functions.range(0, ds.v.num_genotypes()).filter(
+                    lambda oldi: bind(
+                        functions.call(newi),
+                        lambda oldc: functions.gt_index(self.old_to_new[oldc.gtj()],
+                                                        self.old_to_new[oldc.gtk()]) == newi))
+                              .map(lambda oldi: ds.PL[oldi]).min())),
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: would be easier to read with .filter and .map similarly indented, see comment below

File: python/hail/methods/statgen.py [hunk 6]
```diff
@@ -2025,9 +2025,170 @@ def annotate_entries(self, **named_exprs):
...
+        self.annotate_entries(
+            GT = functions.gt_index(self.old_to_new[ds.GT.gtj()],
+                                    self.old_to_new[ds.GT.gtk()]),
+            AD = functions.cond(
+                functions.is_defined(ds.AD),
+                functions.range(0, self.new_v.num_alleles()).map(
+                    lambda newi: functions.range(0, ds.v.num_alleles()).filter(
+                        lambda oldi: self.old_to_new[oldi] == newi).map(
+                            lambda oldi: ds.AD[oldi]).sum()),
+                functions.null(TArray(TInt32()))),
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: I think this style is easier to read in these the PL and AD functions.
```
functions.range(0, self.new_v.num_alleles())
    .map(lambda newi: functions.range(0, ds.v.num_alleles())
        .filter(lambda oldi: self.old_to_new[oldi] == newi)
        .map(lambda oldi: ds.AD[oldi]).sum()),
```

File: python/hail/methods/statgen.py (hunk 3 above)
Comment (cseed) [reply python/hail/methods/statgen.py]: Why?

File: python/hail/methods/statgen.py [hunk 7]
```diff
@@ -2356,9 +2356,200 @@ def annotate_entries(self, **named_exprs):
...
+                functions.range(0, self.new_v.num_alleles()).map(
+                    lambda newi: ds.AD[self.new_to_old[newi]]),
+                functions.null(TArray(TInt32()))),
+            # DP unchanged
+            GQ = functions.gq_from_pl(newPL),
+            PL = newPL)
+
+    def downcode_entries_hts(self):
+        """Use the downcode algorithm to update the matrix table entries for
+        the new, allele filtered variant.
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: allele-filtered

File: python/hail/methods/statgen.py [hunk 8]
```diff
@@ -2356,9 +2356,200 @@ def annotate_entries(self, **named_exprs):
...
+        are added to the depth of the reference allele. Where
+        downcoding filtered alleles merges distinct genotypes, the
+        minimum PL is used (since PL is on a log scale, this roughly
+        corresponds to adding probabilities). The PLs are then
+        re-normalized (shifted) so that the most likely genotype has a
+        PL of 0, and GT is set to this genotype.  If an allele is
+        filtered, this algorithm acts similarly to
+        :func:`.split_multi_hts`.
+
+        The downcoding algorithm would produce the following:
```
Comment (jbloom22) [inline python/hail/methods/statgen.py]: downcode

File: python/hail/methods/statgen.py (hunk 3 above)
Comment (jbloom22) [reply python/hail/methods/statgen.py]: Seemed to be our convention. Up to you.

**Pull request merged**